From 65f17c8f7777d03a64cb1daf47781952e6e1bc5c Mon Sep 17 00:00:00 2001
From: "Yan, WalterX" <walterx.yan@intel.com>
Date: Wed, 11 Jul 2018 18:30:00 +0800
Subject: [PATCH 2/2] Save state and try to restore after MediaPlayer error.

That the mediasever is killed cause MediaPlayer being
unable to restore the state and playing the next one.

Test: see the steps in OAM-65710

Change-Id: I887186c08906a90ef46f1b5a413ed6cfe9ad4dbc
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-65710
Signed-off-by: Yan, WalterX <walterx.yan@intel.com>
---
 .../android/car/media/localmediaplayer/Player.java | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/src/com/android/car/media/localmediaplayer/Player.java b/src/com/android/car/media/localmediaplayer/Player.java
index 0eb65fa..dff49e4 100644
--- a/src/com/android/car/media/localmediaplayer/Player.java
+++ b/src/com/android/car/media/localmediaplayer/Player.java
@@ -28,6 +28,7 @@ import android.media.MediaDescription;
 import android.media.MediaMetadata;
 import android.media.MediaPlayer;
 import android.media.MediaPlayer.OnCompletionListener;
+import android.media.MediaPlayer.OnErrorListener;
 import android.media.session.MediaSession;
 import android.media.session.MediaSession.QueueItem;
 import android.media.session.PlaybackState;
@@ -109,6 +110,7 @@ public class Player extends MediaSession.Callback {
         mMediaPlayer = new MediaPlayer();
         mMediaPlayer.reset();
         mMediaPlayer.setOnCompletionListener(mOnCompletionListener);
+        mMediaPlayer.setOnErrorListener(mOnErrorListener);
         mErrorState = new PlaybackState.Builder()
                 .setState(PlaybackState.STATE_ERROR, 0, 0)
                 .setErrorMessage(context.getString(R.string.playback_error))
@@ -172,6 +174,15 @@ public class Player extends MediaSession.Callback {
         if (Log.isLoggable(TAG, Log.DEBUG)) {
             Log.d(TAG, "onPlay");
         }
+
+        //Here is a specific restore handling as the current media framework mechanism
+        //Only restore state and don't playback when STATE_STOPPED (after onError)
+        if (mSession.getController().getPlaybackState().getState() ==
+                PlaybackState.STATE_STOPPED) {
+            maybeRestoreState();
+            return;
+        }
+
         requestAudioFocus(() -> resumePlayback());
     }
 
@@ -394,6 +405,9 @@ public class Player extends MediaSession.Callback {
             Log.d(TAG, "pausePlayback()");
         }
 
+        //when paused, save state for restoring if necessary
+        saveState();
+
         long currentPosition = 0;
         if (mMediaPlayer.isPlaying()) {
             currentPosition = mMediaPlayer.getCurrentPosition();
@@ -616,4 +630,12 @@ public class Player extends MediaSession.Callback {
             safeAdvance();
         }
     };
+
+    private OnErrorListener mOnErrorListener = new OnErrorListener() {
+        @Override
+        public boolean onError(MediaPlayer mp, int what, int extra) {
+            stopPlayback();
+            return true;
+        }
+    };
 }
-- 
1.9.1

